Update to allow many objects in a compound object

jonathangreen opened the pull request:
**JIRA Ticket**:

https://jira.duraspace.org/browse/ISLANDORA-1846

# What does this Pull Request do?

Currently when a compound object is loaded we call `islandora_object_load` on each member of the compound object twice during the process of rendering the page. This causes page loads to be very slow when a compound object has many children. This pull request eliminates these calls when using the "JAIL" block so that compounds will load fairly quickly.

# What's new?

* Modify the `islandora_compound_object_retrieve_compound_info` function 
  * So it can accept either  the child of a compound object or the parent. Before it would only accept the child, so we had to do checks before we called it to make sure we were the child not the parent. This allows us to eliminate one RI query when loading the object.
  * To return two extra values in its return array:
    * `current_pid`
    * `siblings_detailed`
* Modify the SPARQL query used to find the children of the compound object to add filters for XACML so if an object has XACML on it, it will be filtered out on a per-user basis. This query will work just like the one used before if the repository is not using XACML. This lets us eliminate one use of `islandora_object_load`.
* Modify the javascript code we use to lazy load the images using the JAIL library, so that if an image isn't found at load time we fall back to a default image. This lets us eliminate the other use of `islandora_object_load`.  

# How should this be tested?

* Load a large number of children to a compound object (I've been using 500 in my testing). 
  * Test that before the pull request the page takes a long time to load (minutes). 
  * Test that after applying the pull request the page loads in a normal amount of time.
  * Apply an XACML object to block viewing of an object
    * Make sure that it appears when a user has access. 
    * Make sure it doesn't apply when a user doesn't have access.
  * Apply an XACML policy that blocks an object TN datastream. 
    * Make sure that the JAIL display correctly falls back to the usual islandora "folder" image. 

# Interested parties
@whikloj @jordandukart

jonathangreen committed:
Update to allow many objects in a compound object

When loading a compound object we were doing a islandora_object_load on
each object in the compound object.

This commit improves the performance of the JAIL display block by:
  - Filtering the RI search on XACML predicates to eliminate the need to
    load every item to check access.
  - Checking if an item has a thumbnail that can be accessed when its
    being lazy loaded in javascript instead of on the PHP side.

whikloj commented on islandora_compound_object.module:
```diff
@@ -613,41 +630,53 @@ function islandora_compound_object_retrieve_compound_info(AbstractObject $object
+    // XXX: Handle the case where we are a member of multiple compounds?
     foreach ($part_of as $part) {
       $parent_pid = $part['object']['value'];
       $parent = islandora_object_load($parent_pid);
```
Your comment on line 637 is a good point? This could conceivably load a lot of objects and we only use the last one...right?  If we are going to grab one parent at random, would we be better to just pop the last value off the array and exit quickly?

jonathangreen replied:
Yes. I was just using the behaviour that we had before where we load each possible parent. In practice most times we will only have one parent. 

I submitted a new commit that just directly uses the last value in the array instead of bothering to iterate.

whikloj reviewed (changes requested):
@jonathangreen so I have a couple problems.
1. when I created a compound and restricted viewing one of the objects with XACML it still appears in the compound navigation list. I can't see it (it uses the folder for the thumbnail), but the old system didn't show it at all.
2. If I don't have access to view all the parts of the compound I get the warning.
    `Notice: Trying to get property of non-object in template_preprocess_islandora_compound_prev_next() (line 59 of /var/www/drupal/sites/all/modules/islandora_solution_pack_compound/theme/theme.inc).`
    
    On every page of the compound.

However it is at least twice as fast as the old way when processing a 200 page compound and not using the JAIL display. I'll test that next.

jonathangreen commented:
@whikloj  good catch. Thanks.

I just rebased and pushed a new commit. My bad, there was an issue with the SPARQL query I didn't catch in my testing. Both your issues should be fixed now.

jonathangreen commented:
@whikloj ever get a chance to take another look at this?

whikloj commented:
@jonathangreen sorry got stuck on "work" work. I am good with this, I'm not really sure whom else is using compounds extensively. I'll leave it here with a 👍  and if no one chimes in by this afternoon I'll merge it.

jordandukart commented:
I'll have a look at this afternoon @whikloj and @jonathangreen

whikloj commented on islandora_compound_object.module:
```diff
@@ -415,19 +415,39 @@ function islandora_compound_object_get_parts($pid, $ret_title = FALSE) {
+      ?object islandora-rels-ext:isSequenceNumberOf$escaped_pid ?seq
   }
+  OPTIONAL {{
+      ?object <islandora-rels-ext:isViewableByRole> ?role
```
Just caught this, you are using a prefix defined in the query here. Need to remove the < and > like in line 443.

whikloj commented on islandora_compound_object.module:
```diff
@@ -415,19 +415,39 @@ function islandora_compound_object_get_parts($pid, $ret_title = FALSE) {
+  OPTIONAL {{
+      ?object <islandora-rels-ext:isViewableByRole> ?role
+    } UNION {
+      ?object <islandora-rels-ext:isViewableByUser> ?user
```
Same as above. I am seeing warnings in my Fedora log.
```
WARN 2016-11-30 12:51:32.412 [http-bio-8080-exec-23] (SPARQL) [line: 11, col: 15] Bad IRI: <islandora-rels-ext:isViewableByRole> Code: 45/UNREGISTERED_NONIETF_SCHEME_TREE in SCHEME: The scheme name has a "-" in it, but it does not start in "x-" and the prefix is not known as the prefix of an alternative tree for URI schemes.
```

jonathangreen committed:
Define new hook to supply queries for compound

In order to preserve the old behavior where compound objects load every
item, a new hook hook_islandora_compound_object_query_backends is
provided, where modules can register ways to query compounds. We also
register a new variable islandora_compound_object_query_backend, which
is used to select the backend that will be used.

jonathangreen commented:
Good call. Pushed a commit with those troublesome braces removed. My vagrant is giving me issues, but will test this locally soon.

jonathangreen commented:
Its odd, even with my fedora logs set to DEBUG, I still can't get that same warning you were getting. But my most recent commit with the braces removed works for me and still returns the correct query results for items with XACML in my vagrant box, so it should be all good. Let me know if that fixes your issue.

jordandukart commented on islandora_compound_object.module:
```diff
@@ -450,18 +470,15 @@ EOQ;
 
     foreach ($results as $result) {
-      $object = islandora_object_load($result['object']['value']);
-      if (islandora_object_access(ISLANDORA_VIEW_OBJECTS, $object)) {
```
Not sure this can be removed out right. While XACML falls within the `islandora_object_access` bounds there could be other modules making assertions on viewing.

jonathangreen replied:
Isn't this exactly how [collections](https://github.com/Islandora/islandora_solution_pack_collection/search?utf8=✓&q=islandora_object_access) work though? Put filters on the collection query and don't make any calls to `islandora_object_access`.

jonathangreen replied:
I kind of see a compound object a lot like a collection where we have many children and can query for their availability using solr or sparql. Is there a way to satisfy the access restrictions you are thinking about without loading every object and then calling islandora_object_access on it?

jordandukart replied:
I think it almost needs to exist for legacy purposes because there could be an implementation out there that is employing its own custom access controls. We could likely do something like this:

- Change the code to have the object_access off by default with a `variable_get` and reap all the sweet sweet non-loading benefits.
- Have an update hook that sets the variable mentioned previously to make object_access be on to preserve the previously set behavior on existing installs

jordandukart commented:
Hopefully get a chance to look @ this afternoon @jonathangreen

jonathangreen commented:
Thanks @jordandukart.

I added a variable like you suggested in your review. I also added a hook so that you can register query providers for compound objects. The module registers two providers, one that does the query the same as before, and a new one that does everything in SPARQL. 

I also updated the new SPARQL query so it will respect namespace restrictions.

jonathangreen commented:
Just rebased and pushed a little fix for an issue I found when testing with namespace restrictions.

jonathangreen commented:
Another friendly ping on this. Anyone have a chance to look at these changes?

whikloj commented:
Sorry @jonathangreen, I'll try to get back to this again this week. Working for Fedora right now.

jonathangreen commented:
Thanks @whikloj :+1:
